File: leetcode/JumpGameII.py
from typing import List


class Solution:
    def jump(self, nums: List[int]) -> int:
        """
        Calculates the minimum number of jumps to reach the last index.
        This is solved using a single-pass greedy algorithm.
        """
        n = len(nums)
        if n <= 1:
            return 0

        # `jumps` counts the number of jumps we've made.
        jumps = 0
        # `current_reach` is the farthest index we can get to with the current number of jumps.
        current_reach = 0
        # `farthest_reach` is the maximum index we can reach from any position within the current jump's range.
        farthest_reach = 0

        # We iterate up to n-1 because we don't need to jump from the last element.
        for i in range(n - 1):
            # Update the farthest possible reach from the current position.
            farthest_reach = max(farthest_reach, i + nums[i])

            # If we've reached the end of the current jump's range...
            if i == current_reach:
                # ...we must take another jump.
                jumps += 1
                # The new range is defined by the farthest reach we've found so far.
                current_reach = farthest_reach

        return jumps

File: leetcode/medium/JumpGameII.java
/**
 * Optimized Solution for Jump Game II (Minimum Jumps to Reach the End).
 * * Time Complexity: O(N) - We iterate through the array exactly once.
 * Space Complexity: O(1) - We only use a few integer variables.
 */
class Solution {
    /**
     * Calculates the minimum number of jumps required to reach the last index.
     * * @param nums An array of non-negative integers where nums[i] is the maximum jump length from index i.
     * @return The minimum number of jumps to reach nums.length - 1.
     */
    public int jump(int[] nums) {
        // Base case: If the array has 0 or 1 element, 0 jumps are needed.
        if (nums.length <= 1) {
            return 0;
        }

        int jumps = 0;              // The total minimum number of jumps taken.
        int currentReach = 0;       // The farthest index reachable with the current number of jumps ('jumps').
        int farthestReach = 0;      // The farthest index we can reach *globally* so far, considering all indices up to 'i'.

        // We only need to iterate up to the second-to-last element (nums.length - 2),
        // as reaching currentReach >= nums.length - 1 means we're done.
        for (int i = 0; i < nums.length - 1; i++) {
            
            // 1. Update the overall farthest reachable index (greedy choice).
            // We constantly look for the best potential next jump within the current range.
            farthestReach = Math.max(farthestReach, i + nums[i]);

            // 2. Check if we've reached the limit of the current jump.
            // When 'i' hits 'currentReach', it means we've explored all spots
            // reachable from the *previous* jump and must take a new one.
            if (i == currentReach) {
                // Take the jump.
                jumps++;
                
                // Update the new limit for the next jump.
                // The new 'currentReach' is the 'farthestReach' we found while exploring 
                // the previous jump's range.
                currentReach = farthestReach;
                
                // Early exit: If the new 'currentReach' is already at or past the end,
                // we've found the minimum number of jumps.
                if (currentReach >= nums.length - 1) {
                    return jumps;
                }
            }
        }
        
        // This line handles the case where the loop completes without the early exit 
        // (which shouldn't happen if a solution exists, but is a safe fallback).
        return jumps;          
    }
}